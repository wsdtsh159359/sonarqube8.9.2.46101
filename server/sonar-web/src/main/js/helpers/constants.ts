/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { colors } from '../app/theme';
import { AlmKeys } from '../types/alm-settings';
import { ComponentQualifier } from '../types/component';
import { IssueScope, IssueType } from '../types/issues';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  IssueType.Bug,
  IssueType.Vulnerability,
  IssueType.CodeSmell,
  IssueType.SecurityHotspot
];
export const SOURCE_SCOPES = [
  { scope: IssueScope.Main, qualifier: ComponentQualifier.File },
  { scope: IssueScope.Test, qualifier: ComponentQualifier.TestFile }
];
export const RULE_TYPES: T.RuleType[] = ['BUG', 'VULNERABILITY', 'CODE_SMELL', 'SECURITY_HOTSPOT'];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  colors.red,
  colors.orange,
  colors.yellow,
  colors.lightGreen,
  colors.green
];

export const RATING_COLORS = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const PROJECT_KEY_MAX_LEN = 400;

export const ALM_DOCUMENTATION_PATHS = {
  [AlmKeys.Azure]: '/documentation/analysis/azuredevops-integration/',
  [AlmKeys.BitbucketServer]: '/documentation/analysis/bitbucket-integration/',
  [AlmKeys.BitbucketCloud]: '/documentation/analysis/bitbucket-cloud-integration/',
  [AlmKeys.GitHub]: '/documentation/analysis/github-integration/',
  [AlmKeys.GitLab]: '/documentation/analysis/gitlab-integration/'
};

export const IMPORT_COMPATIBLE_ALMS = [
  AlmKeys.Azure,
  AlmKeys.BitbucketServer,
  AlmKeys.GitHub,
  AlmKeys.GitLab
];
