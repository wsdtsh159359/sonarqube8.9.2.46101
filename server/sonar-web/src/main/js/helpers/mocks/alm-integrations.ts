/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  AzureProject,
  AzureRepository,
  BitbucketProject,
  BitbucketRepository,
  GithubRepository,
  GitlabProject
} from '../../types/alm-integration';

export function mockAzureProject(overrides: Partial<AzureProject> = {}): AzureProject {
  return {
    name: 'azure-project-1',
    description: 'Azure Project',
    ...overrides
  };
}

export function mockAzureRepository(overrides: Partial<AzureRepository> = {}): AzureRepository {
  return {
    name: 'Azure repo 1',
    projectName: 'Azure Project',
    ...overrides
  };
}

export function mockBitbucketProject(overrides: Partial<BitbucketProject> = {}): BitbucketProject {
  return {
    id: 1,
    key: 'project',
    name: 'Project',
    ...overrides
  };
}

export function mockBitbucketRepository(
  overrides: Partial<BitbucketRepository> = {}
): BitbucketRepository {
  return {
    id: 1,
    slug: 'project__repo',
    name: 'Repo',
    projectKey: 'project',
    ...overrides
  };
}

export function mockGitHubRepository(overrides: Partial<GithubRepository> = {}): GithubRepository {
  return {
    id: 'id1234',
    key: 'key3456',
    name: 'repository 1',
    sqProjectKey: '',
    url: 'owner/repo1',
    ...overrides
  };
}

export function mockGitlabProject(overrides: Partial<GitlabProject> = {}): GitlabProject {
  return {
    id: 'id1234',
    name: 'Awesome Project !',
    slug: 'awesome-project-exclamation',
    pathName: 'Company / Best Projects',
    pathSlug: 'company/best-projects',
    sqProjectKey: '',
    url: 'https://gitlab.company.com/best-projects/awesome-project-exclamation',
    ...overrides
  };
}
