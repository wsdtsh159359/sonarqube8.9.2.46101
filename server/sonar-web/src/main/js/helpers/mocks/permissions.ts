/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mockUser } from '../testMocks';

export function mockPermissionGroup(overrides: Partial<T.PermissionGroup> = {}): T.PermissionGroup {
  return {
    name: 'sonar-admins',
    permissions: ['provisioning'],
    ...overrides
  };
}

export function mockPermissionUser(overrides: Partial<T.PermissionUser> = {}): T.PermissionUser {
  return {
    ...mockUser(),
    active: true,
    name: 'johndoe',
    permissions: ['provisioning'],
    ...overrides
  };
}
