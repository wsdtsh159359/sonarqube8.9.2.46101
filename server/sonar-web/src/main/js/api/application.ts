/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { Application, ApplicationPeriod, ApplicationProject } from '../types/application';
import { Visibility } from '../types/component';

export function getApplicationLeak(
  application: string,
  branch?: string
): Promise<ApplicationPeriod[]> {
  return getJSON('/api/applications/show_leak', { application, branch }).then(
    r => r.leaks,
    throwGlobalError
  );
}

export function getApplicationDetails(application: string, branch?: string): Promise<Application> {
  return getJSON('/api/applications/show', { application, branch }).then(
    r => r.application,
    throwGlobalError
  );
}

export function addApplicationBranch(data: {
  application: string;
  branch: string;
  project: string[];
  projectBranch: string[];
}) {
  return post('/api/applications/create_branch', data).catch(throwGlobalError);
}

export function updateApplicationBranch(data: {
  application: string;
  branch: string;
  name: string;
  project: string[];
  projectBranch: string[];
}) {
  return post('/api/applications/update_branch', data).catch(throwGlobalError);
}

export function deleteApplicationBranch(application: string, branch: string) {
  return post('/api/applications/delete_branch', { application, branch }).catch(throwGlobalError);
}

export function getApplicationProjects(data: {
  application: string;
  p?: number;
  ps?: number;
  q?: string;
  selected: string;
}): Promise<{ paging: T.Paging; projects: ApplicationProject[] }> {
  return getJSON('/api/applications/search_projects', data).catch(throwGlobalError);
}

export function addProjectToApplication(application: string, project: string) {
  return post('/api/applications/add_project', { application, project }).catch(throwGlobalError);
}

export function removeProjectFromApplication(application: string, project: string) {
  return post('/api/applications/remove_project', { application, project }).catch(throwGlobalError);
}

export function refreshApplication(key: string) {
  return post('/api/applications/refresh', { key }).catch(throwGlobalError);
}

export function createApplication(
  name: string,
  description: string,
  key: string | undefined,
  visibility: string
): Promise<{
  application: {
    description?: string;
    key: string;
    name: string;
    visibility: Visibility;
  };
}> {
  return postJSON('/api/applications/create', { description, key, name, visibility }).catch(
    throwGlobalError
  );
}

export function deleteApplication(application: string) {
  return post('/api/applications/delete', { application }).catch(throwGlobalError);
}

export function editApplication(application: string, name: string, description: string) {
  return post('/api/applications/update', { name, description, application }).catch(
    throwGlobalError
  );
}
