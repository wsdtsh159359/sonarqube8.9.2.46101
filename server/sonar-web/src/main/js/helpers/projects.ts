/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ProjectKeyValidationResult } from '../types/component';
import { PROJECT_KEY_MAX_LEN } from './constants';

export function validateProjectKey(projectKey: string): ProjectKeyValidationResult {
  // This is the regex used on the backend:
  //   [\p{Alnum}\-_.:]*[\p{Alpha}\-_.:]+[\p{Alnum}\-_.:]*
  // See sonar-core/src/main/java/org/sonar/core/component/ComponentKeys.java
  const regex = /^[\w\-.:]*[a-z\-_.:]+[\w\-.:]*$/i;
  if (projectKey.length === 0) {
    return ProjectKeyValidationResult.Empty;
  } else if (projectKey.length > PROJECT_KEY_MAX_LEN) {
    return ProjectKeyValidationResult.TooLong;
  } else if (regex.test(projectKey)) {
    return ProjectKeyValidationResult.Valid;
  } else {
    return /^[0-9]+$/.test(projectKey)
      ? ProjectKeyValidationResult.OnlyDigits
      : ProjectKeyValidationResult.InvalidChar;
  }
}
