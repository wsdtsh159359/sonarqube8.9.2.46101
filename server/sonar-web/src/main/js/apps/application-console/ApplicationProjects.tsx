/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { find, without } from 'lodash';
import * as React from 'react';
import SelectList, {
  SelectListFilter,
  SelectListSearchParams
} from 'sonar-ui-common/components/controls/SelectList';
import QualifierIcon from 'sonar-ui-common/components/icons/QualifierIcon';
import {
  addProjectToApplication,
  getApplicationProjects,
  removeProjectFromApplication
} from '../../api/application';
import { Application, ApplicationProject } from '../../types/application';

interface Props {
  onAddProject?: (project: ApplicationProject) => void;
  onRemoveProject?: (projectKey: string) => void;
  application: Application;
}

interface State {
  disabledProjects: string[];
  lastSearchParams: SelectListSearchParams & { applicationKey: string };
  needToReload: boolean;
  projects: Array<ApplicationProject>;
  projectsTotalCount?: number;
  selectedProjects: string[];
}

export default class ApplicationProjects extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);

    this.state = {
      disabledProjects: [],
      lastSearchParams: {
        applicationKey: props.application.key,
        query: '',
        filter: SelectListFilter.Selected
      },
      needToReload: false,
      projects: [],
      selectedProjects: []
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.application.key !== this.props.application.key) {
      this.setState(
        prevState => {
          return {
            lastSearchParams: {
              ...prevState.lastSearchParams,
              applicationKey: this.props.application.key
            }
          };
        },
        () => this.fetchProjects(this.state.lastSearchParams)
      );
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  loadApplicationProjects = (searchParams: SelectListSearchParams) =>
    getApplicationProjects({
      application: this.state.lastSearchParams.applicationKey,
      p: searchParams.page,
      ps: searchParams.pageSize,
      q: searchParams.query !== '' ? searchParams.query : undefined,
      selected: searchParams.filter
    });

  fetchProjects = (searchParams: SelectListSearchParams) =>
    this.loadApplicationProjects(searchParams).then(data => {
      if (this.mounted) {
        this.setState(prevState => {
          const more = searchParams.page != null && searchParams.page > 1;

          const { projects, selectedProjects, disabledProjects } = this.dealWithProjects(
            data,
            more,
            prevState
          );

          return {
            disabledProjects,
            lastSearchParams: { ...prevState.lastSearchParams, ...searchParams },
            needToReload: false,
            projects,
            projectsTotalCount: data.paging.total,
            selectedProjects
          };
        });
      }
    });

  dealWithProjects = (
    data: { projects: Array<ApplicationProject>; paging: T.Paging },
    more: boolean,
    prevState: Readonly<State>
  ) => {
    const projects = more ? [...prevState.projects, ...data.projects] : data.projects;

    const newSelectedProjects = data.projects
      .filter(project => project.selected)
      .map(project => project.key);
    const selectedProjects = more
      ? [...prevState.selectedProjects, ...newSelectedProjects]
      : newSelectedProjects;

    const disabledProjects = more ? [...prevState.disabledProjects] : [];

    return {
      disabledProjects,
      projects,
      selectedProjects
    };
  };

  handleSelect = (projectKey: string) => {
    return addProjectToApplication(this.props.application.key, projectKey).then(() => {
      if (this.mounted) {
        this.setState(state => {
          const project = state.projects.find(p => p.key === projectKey);
          if (project && this.props.onAddProject) {
            this.props.onAddProject(project);
          }
          return {
            needToReload: true,
            selectedProjects: [...state.selectedProjects, projectKey]
          };
        });
      }
    });
  };

  handleUnselect = (projectKey: string) => {
    return removeProjectFromApplication(this.props.application.key, projectKey).then(() => {
      if (this.mounted) {
        this.setState(state => {
          if (this.props.onRemoveProject) {
            this.props.onRemoveProject(projectKey);
          }
          return {
            needToReload: true,
            selectedProjects: without(state.selectedProjects, projectKey)
          };
        });
      }
    });
  };

  renderElement = (projectKey: string) => {
    const project = find(this.state.projects, { key: projectKey });
    if (project === undefined) {
      return '';
    }

    return (
      <div className="views-project-item display-flex-center">
        <QualifierIcon className="spacer-right" qualifier="TRK" />
        <div>
          <div title={project.name}>{project.name}</div>
          <div className="note">{project.key}</div>
        </div>
      </div>
    );
  };

  render() {
    const { projects, selectedProjects } = this.state;

    return (
      <SelectList
        disabledElements={this.state.disabledProjects}
        elements={projects.map(project => project.key)}
        elementsTotalCount={this.state.projectsTotalCount}
        needToReload={
          this.state.needToReload &&
          this.state.lastSearchParams &&
          this.state.lastSearchParams.filter !== SelectListFilter.All
        }
        onSearch={this.fetchProjects}
        onSelect={this.handleSelect}
        onUnselect={this.handleUnselect}
        renderElement={this.renderElement}
        selectedElements={selectedProjects}
        withPaging={true}
      />
    );
  }
}
