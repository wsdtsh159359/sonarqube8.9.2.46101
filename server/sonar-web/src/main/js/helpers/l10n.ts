/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { toNotSoISOString } from 'sonar-ui-common/helpers/dates';
import SonarUiCommonInitializer, { DEFAULT_LOCALE } from 'sonar-ui-common/helpers/init';
import {
  get as loadFromLocalStorage,
  save as saveInLocalStorage
} from 'sonar-ui-common/helpers/storage';
import { fetchL10nBundle } from '../api/l10n';
import { L10nBundle, L10nBundleRequestParams } from '../types/l10n';

const L10N_BUNDLE_LS_KEY = 'l10n.bundle';

export async function loadL10nBundle() {
  const bundle = await getLatestL10nBundle().catch(() => ({
    locale: DEFAULT_LOCALE,
    messages: {}
  }));

  SonarUiCommonInitializer.setLocale(bundle.locale).setMessages(bundle.messages);
  // No need to load english (default) bundle, it's coming with react-intl
  if (bundle.locale !== DEFAULT_LOCALE) {
    const [intlBundle, intl] = await Promise.all([
      import(`react-intl/locale-data/${bundle.locale}`),
      import('react-intl')
    ]);

    intl.addLocaleData(intlBundle.default);
  }

  return bundle;
}

export async function getLatestL10nBundle() {
  const browserLocale = getPreferredLanguage();
  const cachedBundle = loadL10nBundleFromLocalStorage();

  const params: L10nBundleRequestParams = {};

  if (browserLocale) {
    params.locale = browserLocale;

    if (
      cachedBundle.locale &&
      browserLocale.startsWith(cachedBundle.locale) &&
      cachedBundle.timestamp &&
      cachedBundle.messages
    ) {
      params.ts = cachedBundle.timestamp;
    }
  }

  const { effectiveLocale, messages } = await fetchL10nBundle(params).catch(response => {
    if (response && response.status === 304) {
      return {
        effectiveLocale: cachedBundle.locale || browserLocale || DEFAULT_LOCALE,
        messages: cachedBundle.messages ?? {}
      };
    } else {
      throw new Error(`Unexpected status code: ${response.status}`);
    }
  });

  const bundle = {
    timestamp: toNotSoISOString(new Date()),
    locale: effectiveLocale,
    messages
  };

  saveL10nBundleToLocalStorage(bundle);

  return bundle;
}

export function getCurrentL10nBundle() {
  return loadL10nBundleFromLocalStorage();
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function loadL10nBundleFromLocalStorage() {
  let bundle: L10nBundle;

  try {
    bundle = JSON.parse(loadFromLocalStorage(L10N_BUNDLE_LS_KEY) ?? '{}');
  } catch {
    bundle = {};
  }

  return bundle;
}

function saveL10nBundleToLocalStorage(bundle: L10nBundle) {
  saveInLocalStorage(L10N_BUNDLE_LS_KEY, JSON.stringify(bundle));
}
