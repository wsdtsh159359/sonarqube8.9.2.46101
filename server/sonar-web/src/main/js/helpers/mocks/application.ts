/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Application, ApplicationPeriod, ApplicationProject } from '../../types/application';
import { Visibility } from '../../types/component';
import { mockBranch } from './branch-like';

export function mockApplication(overrides: Partial<Application> = {}): Application {
  return {
    branches: [mockBranch()],
    key: 'foo',
    name: 'Foo',
    projects: [mockApplicationProject()],
    visibility: Visibility.Private,
    ...overrides
  };
}

export function mockApplicationPeriod(
  overrides: Partial<ApplicationPeriod> = {}
): ApplicationPeriod {
  return {
    date: '2017-10-01',
    project: 'foo',
    projectName: 'Foo',
    ...overrides
  };
}

export function mockApplicationProject(
  overrides: Partial<ApplicationProject> = {}
): ApplicationProject {
  return {
    branch: 'master',
    isMain: true,
    key: 'bar',
    name: 'Bar',
    ...overrides
  };
}
