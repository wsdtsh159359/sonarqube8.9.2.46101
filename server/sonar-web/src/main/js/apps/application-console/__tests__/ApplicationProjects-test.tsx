/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import SelectList, { SelectListFilter } from 'sonar-ui-common/components/controls/SelectList';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import {
  addProjectToApplication,
  getApplicationProjects,
  removeProjectFromApplication
} from '../../../api/application';
import { mockApplication } from '../../../helpers/mocks/application';
import ApplicationProjects from '../ApplicationProjects';

jest.mock('../../../api/application', () => ({
  getApplicationProjects: jest.fn().mockResolvedValue({
    paging: { pageIndex: 1, pageSize: 3, total: 55 },
    projects: [
      { key: 'test1', name: 'test1', selected: false },
      { key: 'test2', name: 'test2', selected: false, disabled: true, includedIn: 'foo' },
      { key: 'test3', name: 'test3', selected: true }
    ]
  }),
  addProjectToApplication: jest.fn().mockResolvedValue({}),
  removeProjectFromApplication: jest.fn().mockResolvedValue({})
}));

beforeEach(jest.clearAllMocks);

it('should render correctly in application mode', async () => {
  const wrapper = shallowRender();
  wrapper
    .find(SelectList)
    .props()
    .onSearch({ query: '', filter: SelectListFilter.Selected, page: 1, pageSize: 100 });
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.instance().renderElement('test1')).toMatchSnapshot();
  expect(wrapper.instance().renderElement('test2')).toMatchSnapshot();

  expect(getApplicationProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      application: 'foo',
      p: 1,
      ps: 100,
      q: undefined,
      selected: SelectListFilter.Selected
    })
  );

  wrapper.instance().handleSelect('test1');
  await waitAndUpdate(wrapper);
  expect(addProjectToApplication).toHaveBeenCalledWith('foo', 'test1');

  wrapper.instance().fetchProjects({ query: 'bar', filter: SelectListFilter.Selected });
  await waitAndUpdate(wrapper);
  expect(getApplicationProjects).toHaveBeenCalledWith(
    expect.objectContaining({ application: 'foo', q: 'bar', selected: SelectListFilter.Selected })
  );

  wrapper.instance().handleUnselect('test1');
  await waitAndUpdate(wrapper);
  expect(removeProjectFromApplication).toHaveBeenCalledWith('foo', 'test1');
});

it('should refresh properly if props changes', () => {
  const wrapper = shallowRender();
  const spy = jest.spyOn(wrapper.instance(), 'fetchProjects');

  wrapper.setProps({ application: { key: 'bar' } as any });
  expect(wrapper.state().lastSearchParams.applicationKey).toBe('bar');
  expect(spy).toHaveBeenCalled();
});

function shallowRender(props: Partial<ApplicationProjects['props']> = {}) {
  return shallow<ApplicationProjects>(
    <ApplicationProjects application={mockApplication()} {...props} />
  );
}
