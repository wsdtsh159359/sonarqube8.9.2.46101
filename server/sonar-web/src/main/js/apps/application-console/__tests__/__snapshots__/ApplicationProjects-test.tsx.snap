// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly in application mode 1`] = `
<SelectList
  disabledElements={Array []}
  elements={
    Array [
      "test1",
      "test2",
      "test3",
    ]
  }
  elementsTotalCount={55}
  needToReload={false}
  onSearch={[Function]}
  onSelect={[Function]}
  onUnselect={[Function]}
  renderElement={[Function]}
  selectedElements={
    Array [
      "test3",
    ]
  }
  withPaging={true}
/>
`;

exports[`should render correctly in application mode 2`] = `
<div
  className="views-project-item display-flex-center"
>
  <QualifierIcon
    className="spacer-right"
    qualifier="TRK"
  />
  <div>
    <div
      title="test1"
    >
      test1
    </div>
    <div
      className="note"
    >
      test1
    </div>
  </div>
</div>
`;

exports[`should render correctly in application mode 3`] = `
<div
  className="views-project-item display-flex-center"
>
  <QualifierIcon
    className="spacer-right"
    qualifier="TRK"
  />
  <div>
    <div
      title="test2"
    >
      test2
    </div>
    <div
      className="note"
    >
      test2
    </div>
  </div>
</div>
`;
